Systemd optional deps

software-artificer opened the pull request:
###### Description of changes
Allow to slim down the systemd build by:

- Making ACL (access control list) support via `libacl` optional;
- Making Linux Audit Framework integration optional;
- Making kmod/modprobe integration optional;
- Making IDN (Internationalized Domain Names) support via `libidn2` optional;
- Making PAM (Pluggable Authentication Module) stack integration optional;
- Fix the `systemd-initrd-simple` test

Fixes #216461 

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] ~For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))~ Testing on NixOS.
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).


###### Test suites
Tests were executed from the current `master` branch with my commits cherry-picked on top in order to avoid some broken state in the current `staging`.

Test logs are uploaded to pastebin due to their massive size and GH limits
- [[root@nixos:~/nixpkgs]# nix-build -A nixosTests.hibernate-systemd-stage-1](https://github.com/filakhtov/test-proof-artifacts/blob/bd2dd505ee2301148289c8d4383e1c661b152a19/gh/nixos/nixpkgs/217249/hibernate-systemd-stage-1.log)
- [[root@nixos:~/nixpkgs]# nix-build -A nixosTests.initrdNetwork](https://github.com/filakhtov/test-proof-artifacts/blob/bd2dd505ee2301148289c8d4383e1c661b152a19/gh/nixos/nixpkgs/217249/initrdNetwork.log)
- [[root@nixos:~/nixpkgs]# nix-build -A nixosTests.installer.bcache](https://github.com/filakhtov/test-proof-artifacts/blob/bd2dd505ee2301148289c8d4383e1c661b152a19/gh/nixos/nixpkgs/217249/installer.bcache.log)
- [[root@nixos:~/nixpkgs]# nix-build -A nixosTests.installer.bcachefsEncrypted](https://github.com/filakhtov/test-proof-artifacts/blob/bd2dd505ee2301148289c8d4383e1c661b152a19/gh/nixos/nixpkgs/217249/installer.bcachefsEncrypted.log)
- [[root@nixos:~/nixpkgs]# nix-build -A nixosTests.installer.bcachefsMulti](https://github.com/filakhtov/test-proof-artifacts/blob/bd2dd505ee2301148289c8d4383e1c661b152a19/gh/nixos/nixpkgs/217249/installer.bcachefsMulti.log)
- [[root@nixos:~/nixpkgs]# nix-build -A nixosTests.installer.bcachefsSimple](https://github.com/filakhtov/test-proof-artifacts/blob/bd2dd505ee2301148289c8d4383e1c661b152a19/gh/nixos/nixpkgs/217249/installer.bcachefsSimple.log)
- [[root@nixos:~/nixpkgs]# nix-build -A nixosTests.installer.btrfsSimple](https://github.com/filakhtov/test-proof-artifacts/blob/bd2dd505ee2301148289c8d4383e1c661b152a19/gh/nixos/nixpkgs/217249/installer.btrfsSimple.log)
- [[root@nixos:~/nixpkgs]# nix-build -A nixosTests.installer.btrfsSubvolDefault](https://github.com/filakhtov/test-proof-artifacts/blob/bd2dd505ee2301148289c8d4383e1c661b152a19/gh/nixos/nixpkgs/217249/installer.btrfsSubvolDefault.log)
- [[root@nixos:~/nixpkgs]# nix-build -A nixosTests.installer.btrfsSubvolEscape](https://github.com/filakhtov/test-proof-artifacts/blob/bd2dd505ee2301148289c8d4383e1c661b152a19/gh/nixos/nixpkgs/217249/installer.btrfsSubvolEscape.log)
- [[root@nixos:~/nixpkgs]# nix-build -A nixosTests.installer.btrfsSubvols](https://github.com/filakhtov/test-proof-artifacts/blob/bd2dd505ee2301148289c8d4383e1c661b152a19/gh/nixos/nixpkgs/217249/installer.btrfsSubvols.log)
- [[root@nixos:~/nixpkgs]# nix-build -A nixosTests.installer.encryptedFSWithKeyfile](https://github.com/filakhtov/test-proof-artifacts/blob/bd2dd505ee2301148289c8d4383e1c661b152a19/gh/nixos/nixpkgs/217249/installer.encryptedFSWithKeyfile.log)
- [[root@nixos:~/nixpkgs]# nix-build -A nixosTests.installer.fullDiskEncryption](https://github.com/filakhtov/test-proof-artifacts/blob/bd2dd505ee2301148289c8d4383e1c661b152a19/gh/nixos/nixpkgs/217249/installer.fullDiskEncryption.log)
- [[root@nixos:~/nixpkgs]# nix-build -A nixosTests.systemd](https://github.com/filakhtov/test-proof-artifacts/blob/bd2dd505ee2301148289c8d4383e1c661b152a19/gh/nixos/nixpkgs/217249/systemd.log)
- [[root@nixos:~/nixpkgs]# nix-build -A nixosTests.systemd-initrd-btrfs-raid](https://github.com/filakhtov/test-proof-artifacts/blob/bd2dd505ee2301148289c8d4383e1c661b152a19/gh/nixos/nixpkgs/217249/systemd-initrd-btrfs-raid.log)
- [[root@nixos:~/nixpkgs]# nix-build -A nixosTests.systemd-initrd-luks-fido2](https://github.com/filakhtov/test-proof-artifacts/blob/bd2dd505ee2301148289c8d4383e1c661b152a19/gh/nixos/nixpkgs/217249/systemd-initrd-luks-fido2.log)
- [[root@nixos:~/nixpkgs]# nix-build -A nixosTests.systemd-initrd-luks-keyfile](https://github.com/filakhtov/test-proof-artifacts/blob/bd2dd505ee2301148289c8d4383e1c661b152a19/gh/nixos/nixpkgs/217249/systemd-initrd-luks-keyfile.log)
- [[root@nixos:~/nixpkgs]# nix-build -A nixosTests.systemd-initrd-luks-password](https://github.com/filakhtov/test-proof-artifacts/blob/bd2dd505ee2301148289c8d4383e1c661b152a19/gh/nixos/nixpkgs/217249/systemd-initrd-luks-password.log)
- [[root@nixos:~/nixpkgs]# nix-build -A nixosTests.systemd-initrd-luks-tpm2](https://github.com/filakhtov/test-proof-artifacts/blob/bd2dd505ee2301148289c8d4383e1c661b152a19/gh/nixos/nixpkgs/217249/systemd-initrd-luks-tpm2.log)
- [[root@nixos:~/nixpkgs]# nix-build -A nixosTests.systemd-initrd-modprobe](https://github.com/filakhtov/test-proof-artifacts/blob/bd2dd505ee2301148289c8d4383e1c661b152a19/gh/nixos/nixpkgs/217249/systemd-initrd-modprobe.log)
- [[root@nixos:~/nixpkgs]# nix-build -A nixosTests.systemd-initrd-shutdown](https://github.com/filakhtov/test-proof-artifacts/blob/bd2dd505ee2301148289c8d4383e1c661b152a19/gh/nixos/nixpkgs/217249/systemd-initrd-shutdown.log)
- [[root@nixos:~/nixpkgs]# nix-build -A nixosTests.systemd-initrd-simple](https://github.com/filakhtov/test-proof-artifacts/blob/bd2dd505ee2301148289c8d4383e1c661b152a19/gh/nixos/nixpkgs/217249/systemd-initrd-simple.log)
- [[root@nixos:~/nixpkgs]# nix-build -A nixosTests.systemd-initrd-swraid](https://github.com/filakhtov/test-proof-artifacts/blob/bd2dd505ee2301148289c8d4383e1c661b152a19/gh/nixos/nixpkgs/217249/systemd-initrd-swraid.log)
- [[root@nixos:~/nixpkgs]# nix-build -A nixosTests.systemd-initrd-vconsole](https://github.com/filakhtov/test-proof-artifacts/blob/bd2dd505ee2301148289c8d4383e1c661b152a19/gh/nixos/nixpkgs/217249/systemd-initrd-vconsole.log)

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

software-artificer committed:
systemd: make libidn2 optional

Expose a new `withLibidn2` flag (defauts to true for backwards compatibility) to be able to conditionally enable and disable integration with `libidn2`, which is used by the `systemd-network` and `systemd-resolved` to support internationalized domain names.

software-artificer committed:
systemd: make libacl optional

Expose a new `withAcl` flag (defaults to true for backwards compatibility) to be able to conditionally enable and disable an integration with `libacl` library, which is used by variety of systemd tools and daemon, e.g. `journald` will check ACLs in addition to regular permissions when accessing journal files and `systemd-nspawn` will update ACL entries when used with the `--private-users-chown` flag.

flokli commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -82,6 +82,7 @@
 , bpftools
 , libbpf
 
+, withAcl ? true
```
Can you update the `systemdMinimal` definition (in `pkgs/top-level/all-packages.nix`) to also set these two new options to false if possible?

flokli replied:
Note we want to enable idn2 again for `systemdStage1Network`.

flokli reviewed (commented):
LGTM. CI error might be unrelated, and fix itself after another rebase to staging.

software-artificer committed:
systemd: make libaudit optional

Expose a new `withAudit` flag (defaults to `true` for backwards compatibility) to be able to conditionally enable and disable an integration with the `libaudit` library, which is used to integrate with Linux Audit Framework for logging various security-relevant events.

software-artificer committed:
systemd: make PAM integration optional

Expose a new `withPam` option to allow enabling and disabling
integration with PAM stack, including the `systemd-user-sessions` daemon
and the associated `.service` file, as well as `pam_systemd.so` PAM
module for integration with `systemd-logind` and user session
registration with the systemd cgroup hierarchy.

filakhtov commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -129,6 +133,7 @@
 assert withImportd -> withCompression;
 assert withCoredump -> withCompression;
 assert withHomed -> withCryptsetup;
+assert withHomed -> withPam;
```
I have a question on this one: what would be the preferred way to do something like that? Would it be what I have done in the change or something like this instead:
```suggestion
assert withHomed -> withCryptsetup && withPam;
```

software-artificer committed:
systemd: optional kmod integration

Expose a new `withKmod` option to be able to enable and disable kmod
integration, including the `systemd-modules-load` tool for automatic
modules loading during the system boot sequence.

filakhtov commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -711,7 +723,7 @@ stdenv.mkDerivation (finalAttrs: {
     interfaceVersion = 2;
 
-    inherit withCryptsetup withHostnamed withImportd withLocaled withMachined withPortabled withTimedated withUtmp util-linux kmod kbd;
+    inherit withCryptsetup withHostnamed withImportd withKmod withLocaled withMachined withPortabled withTimedated withUtmp util-linux kmod kbd;
```
I am not sure if this is necessary, but it seems that way, because if we go from kmod enable to kmod disabled or vice-versa we might have issues reloading some of the systemd units on `nixos-rebuild switch`. Correct me if I misunderstand this.

software-artificer committed:
systemd: disable dependencies for minimal build

The `systemdMinimal` build is used purely for Udev and therefore does
not require all the extra dependencies that are needed for normal
operation. As more flags were exposed to allow disabling additional
opional dependencies the `systemdMinimal` will now take advantage of
those.

filakhtov commented on pkgs/top-level/all-packages.nix:
```diff
@@ -26702,14 +26704,17 @@ with pkgs;
     withHomed = false;
     withHwdb = false;
     withImportd = false;
+    withKmod = false;
```
As far as I understand we only care about the `libudev` provided by the `systemdMinimal`, so I have disabled the `kmod` integration here, because AFAIK it is only used by the `systemd-modules-load`, but not in the Udev implementation ([this](https://github.com/systemd/systemd/blob/main/src/udev/udev-builtin-kmod.c) is used instead).

lheckemann replied:
mobile-nixos at least uses the udevd from systemdMinimal, so this change breaks mobile-nixos.

I think it's fair enough to say that systemdStage1 or a full systemd should be used instead, but in that case we should probably also prevent udevd and rules from being installed at all in the minimal (libudev-only) variant of the package -- because I imagine cases where a kmod-free `udevd` is indeed desired should be extremely rare and if someone needs that they'll probably know it.

lheckemann replied:
(the kmod-builtin thing is only included if HAVE_KMOD is true https://github.com/systemd/systemd/blob/4f44d2c4f76922a4f48dd4473e6abaca40d7e555/src/udev/meson.build#L39 )

filakhtov replied:
@lheckemann thanks for the information. I will drop a PR later to but it back on. Do we have any tests covering mobile-nixos use case?

filakhtov replied:
Opened #224421, apologies for the inconvenience. A question I have on this: would it make sense to have `systemdMobile` with precisely tailored set of options enabled in order to avoid running into similar problems in the future?

filakhtov commented on pkgs/top-level/all-packages.nix:
```diff
@@ -26722,13 +26727,17 @@ with pkgs;
   systemdStage1Network = systemdStage1.override {
     pname = "systemd-stage-1-network";
     withNetworkd = true;
+    withLibidn2 = true;
```
`libidn2` only makes sense if the networking functions are enabled (resolved or networkd), so I have put it into the `systemdStage1Network` instead of the default `systemdStage1`. Is that reasonable?

flokli replied:
I think so, yes.

filakhtov commented on pkgs/top-level/all-packages.nix:
```diff
@@ -26722,13 +26727,17 @@ with pkgs;
   };
   systemdStage1 = systemdMinimal.override {
     pname = "systemd-stage-1";
+    withAcl = true;
     withCryptsetup = true;
     withFido2 = true;
+    withKmod = true;
+    withPam = true;
```
I assume we care about ACL and kmod in stage1, but I am not sure about the PAM stack. Can it be dropped?

flokli replied:
I'm not sure. It's probably needed to `ssh` in (?), as well as to enter the root password in case there was an error? @ElvishJerricco, any thoughts?

ElvishJerricco replied:
To be honest I'm not entirely sure. We shouldn't need PAM, as we don't include anything else related to PAM (for instance, the ssh module in #169116 has `UsePAM no`). I would wager that kmod is needed for `systemd-modules-load`, in which case yes that is a requirement for systemd stage 1. But I'm not sure about `withAcl`.

filakhtov replied:
Thanks @ElvishJerricco!

If we aren't using PAM this early, I am happy to drop it and re-test relevant things.

ACL I think is needed. I am not deeply familiar with NixOS yet (I have just recently started playing around with it), but a lot of modern system use ACL to allow access to `/dev/dri/` device nodes, and that's something Plymouth for example takes advantage of IIRC.

filakhtov commented:
Hey @flokli, thanks a lot for your guidance so far. I think I have trimmed everything to the bare minimum at this point, but I have a bunch of questions that I would need your guidance on, whenever you have some time 🙏🏻 Thanks in advance.

What would be the best way to test all of this? Do I need to write some tests, and if so what do we want to actually test here? I really haven't had an opportunity to figure out testing just yet, so appreciate any help on this.

flokli commented:
> What would be the best way to test all of this? Do I need to write some tests, and if so what do we want to actually test here? I really haven't had an opportunity to figure out testing just yet, so appreciate any help on this.

We have a bunch of tests in `nixos/tests`, which you can simply run by `nix-build`'ing them. Maybe a good way to start would be the `systemd-initrd-*` ones (make sure to confirm they did succeed before). That should give you an idea if there's any major regressions.

filakhtov commented:
Just a small progress update: I am continuing to work on this, just having some trouble running tests and verifying everything (due to my LXC set-up), so needed to build a VM and a bit of infra to be able to test this properly and will come back with the results at some point soon-ish, time permitting.

filakhtov commented:
So I've got a system running NixOS now, so I can run the tests, but I am having a problem building the test dependencies to be able to run the tests:
```
=================================== FAILURES ===================================
_________________________ TestLexers.testIPythonLexer __________________________

self = <IPython.lib.tests.test_lexers.TestLexers testMethod=testIPythonLexer>

    def testIPythonLexer(self):
        fragment = '!echo $HOME\n'
        tokens = [
            (Token.Operator, '!'),
        ]
        tokens.extend(self.bash_lexer.get_tokens(fragment[1:]))
>       self.assertEqual(tokens, list(self.lexer.get_tokens(fragment)))
E       AssertionError: Lists differ: [(Tok[78 chars] (Token.Name.Variable, '$HOME'), (Token.Text.Whitespace, '\n')] != [(Tok[78 chars] (Token.Name.Variable, '$HOME'), (Token.Text, '\n')]
E       
E       First differing element 4:
E       (Token.Text.Whitespace, '\n')
E       (Token.Text, '\n')
E       
E         [(Token.Operator, '!'),
E          (Token.Name.Builtin, 'echo'),
E          (Token.Text.Whitespace, ' '),
E          (Token.Name.Variable, '$HOME'),
E       -  (Token.Text.Whitespace, '\n')]
E       ?             -----------
E       
E       +  (Token.Text, '\n')]

IPython/lib/tests/test_lexers.py:25: AssertionError
```
Which seems to be a problem between [pygments 2.14.0 and ipython <8.8](https://groups.google.com/g/linux.debian.bugs.dist/c/g1egcKK8p3E).

How would one normally deal with this? Just disable the offending tests for now as they're completely unrelated to the change? Or do I just cherry-pick my change on top of the current stable channel (so I can use prebuilt dependencies) and test thing that way? Any ideas @flokli?

flokli commented:
Sorry for the delay in replying. I think you've been bitten by another unrelated regression in staging, which might be very unstable sometimes (that's what we use the staging-next stabilization cycles for).

In these cases, it's probably easiest to *test* your changes by cherry-picking these commits on top of master (pushed to a separate branch, but still keep the PR open as-is), and once you're happy with the results, merge this PR.

filakhtov commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -661,7 +673,7 @@ stdenv.mkDerivation (finalAttrs: {
 
   postInstall = ''
     mkdir -p $out/example/systemd
-    mv $out/lib/{modules-load.d,binfmt.d,sysctl.d,tmpfiles.d} $out/example
+    mv $out/lib/{binfmt.d,sysctl.d,tmpfiles.d} $out/example
```
Missed this one initially, which was failing the build. Now fixed.

filakhtov commented on nixos/tests/systemd-initrd-simple.nix:
```diff
@@ -7,7 +7,7 @@ import ./make-test-python.nix ({ lib, pkgs, ... }: {
     };
     fileSystems = lib.mkVMOverride {
-      "/".autoResize = true;
+      "/".autoResize = lib.mkForce true;
```
This test was broken and was failing due to:
```
[root@nixos:~/nixpkgs]# nix-build -A nixosTests.systemd-initrd-simple
error: The option `nodes.machine.fileSystems."/".autoResize' has conflicting definition values:
       - In `makeTest parameters': true
       - In `/root/nixpkgs/nixos/modules/virtualisation/qemu-vm.nix': false
       Use `lib.mkForce value` or `lib.mkDefault value` to change the priority on any of these definitions.
(use '--show-trace' to show detailed location information)
```

filakhtov replied:
Subsequently fixed in NixOS/nixpkgs#219716, so I have dropped the last commit.

filakhtov commented:
Hey @flokli, I think I have reached the point where we can start getting ready to merge this. Do you want me to test or verify anything else before? Or do I make this PR ready for review and we work from there?

flokli commented:
@filakhtov thanks! Yes, this looks pretty good now. I think we should be running some more tests (installer tests, systemd-initrd* tests), to make sure there's no new test failures (some of these might already be broken currently).

If there's no new regressions, fine for me to merge this :-)

flokli reviewed (approved):
LGTM. Thanks for the work!

software-artificer committed:
systemdStage1: disable PAM

When PAM was made optional initially, we decided to keep it enabled for stage 1, but as was later pointed out during the code review it is unnecessary, because we never use PAM in stage 1, even in network-enabled stage 1 with OpenSSH we have `UsePAM` set to `no`, so disabling it now.

filakhtov commented:
#### Testing with `withPam = false` in `systemdStage1`
- [[root@nixos:~/nixpkgs]# nix-build -A nixosTests.installer-systemd-stage-1.simple](https://gist.github.com/filakhtov/31248eb672525eca31fa3c047426a1f0#file-installer-systemd-stage-1-simple)

flokli commented:
Let's get this in, and deal with any further regressions during the next stabilization cycle. This is probably better tested than some other staging changes ;-)

Thanks a lot for the work!
